forte/processors/data_augment/algorithms/skeleton_op.py

```diff
@@ -0,0 +1,765 @@
+# Copyright 202 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+SkeletonOp is the most low level Op used for Data Augmentation.
+Through this Op, users can use in-built utility functions to implement
+their own augmentation logic.
+"""
+from collections import defaultdict
+from copy import deepcopy
+from typing import (
+    DefaultDict,
+    Iterable,
+    List,
+    Tuple,
+    Union,
+    Dict,
+    Any,
+    Set,
+    cast,
+)
+from abc import abstractmethod, ABC
+from bisect import bisect_left, bisect_right
+from sortedcontainers.sorteddict import SortedDict
+from sortedcontainers.sortedlist import SortedList
+from forte.data.data_pack import DataPack
+from forte.common.configuration import Config
+from forte.data.span import Span
+from forte.data.ontology.top import (
+    Link,
+    Group,
+    MultiPackGroup,
+    MultiPackLink,
+    Annotation,
+)
+from forte.data.multi_pack import MultiPack
+from forte.data.ontology.core import BaseLink, Entry
+from forte.utils.utils import create_class_with_kwargs, get_class
+
+
+__all__ = ["SkeletonOp"]
+
+
+class SkeletonOp(ABC):
+    r"""
+    The SkeletonOp is the most basic augmentation Op that
+    gives users the most amount of freedom to implement their
+    logic of augmentation. The users are expected to use the
+    provided utility functions to implement their own augmentation
+    logic which will then be substantiated into new data packs. This
+    Op requires the users to have a relatively stronger understanding of
+    Forte's internal setup.
+    """
+
+    def __init__(self, configs: Union[Config, Dict[str, Any]]) -> None:
+
+        # :attr:`_replaced_annos`: {datapack id: SortedList[span, new text]}
+        # It records the spans replaced by new texts.
+        # It is a map from datapack id to a list of tuples
+        # (span, new text) inserted by :func:`replace`.
+        # The new text will be used for building new data pack.
```

Comment:
This sentence `The new text will be used for building new data pack` is slightly not clear.